plugins/hls-class-plugin/src/Ide/Plugin/Class.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE DeriveAnyClass    #-}
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards   #-}
+{-# LANGUAGE TypeFamilies      #-}
+{-# LANGUAGE ViewPatterns      #-}
+module Ide.Plugin.Class
+  ( descriptor
+  ) where
+
+import           BooleanFormula
+import           Class
+import           ConLike
+import           Control.Applicative
+import           Control.Lens hiding (List)
+import           Control.Monad
+import           Data.Aeson
+import           Data.Char
+import           Data.List
+import qualified Data.Map.Strict as Map
+import           Data.Maybe
+import qualified Data.Text as T
+import           Development.IDE
+import           Development.IDE.GHC.Compat
+import           Development.IDE.Spans.AtPoint
+import qualified GHC.Generics as Generics
+import           GhcPlugins hiding (Var, (<>))
+import           Ide.Plugin
+import           Ide.Types
+import           Language.Haskell.LSP.Types
+import qualified Language.Haskell.LSP.Types.Lens as J
+import           TcEnv
+import           TcRnMonad
+import qualified Data.HashMap.Strict as H
+
+descriptor :: PluginId -> PluginDescriptor
+descriptor plId = (defaultPluginDescriptor plId)
+  { pluginCommands = commands
+  , pluginCodeActionProvider = Just codeAction
+  }
+
+commands :: [PluginCommand]
+commands = [ PluginCommand "addMethodPlaceholders" "add placeholders for minimal methods" addMethodPlaceholders
+           ]
+
+-- | Parameter for the addMethods PluginCommand.
+data AddMethodsParams = AddMethodsParams
+  { uri         :: Uri
+  , range       :: Range
+  , methodGroup :: List T.Text
+  }
+  deriving (Show, Eq, Generics.Generic, ToJSON, FromJSON)
+
+addMethodPlaceholders :: CommandFunction AddMethodsParams
+addMethodPlaceholders _ _ AddMethodsParams{..} = pure (Right Null, Just (WorkspaceApplyEdit, ApplyWorkspaceEditParams workspaceEdit))
+  where
+    workspaceEdit
+      = WorkspaceEdit
+        (Just (H.singleton uri textEdits))
+        Nothing
+
+    textEdits
+      = List
+        [ TextEdit (Range pos pos) $ "\n" <> methodText
```

Comment:
This edit is somewhat problematic because it just appends method texts, without checking whether there is `where` keyword or not. For example,
```Haskell
instance Eq X where
```
works well with this plugin, and will produce one of the
```haskell
instance Eq X where
  (==) = _
```
or
```haskell
instance Eq X where
  (/=) = _
```
However, if there is no `where` clause, I believe GHC gives the same error, so
```Haskell
instance Eq X
```
becomes
```Haskell
instance Eq X
  (==) = _
```
with this plugin. This is surely not ideal.

Reply:
I would use ghc-exactprint to create the edit instead of text modifications. 

https://github.com/haskell/haskell-language-server/blob/master/plugins/tactics/src/Ide/TreeTransform.hs has a nice collection of utilities that make that very feasible

Reply:
Great, I will check that.